### PR TITLE
Create a validate free function in zim ns

### PR DESCRIPTION
**Pull request by MiguelRocha**
This PR introduces the first step for libzim validation. It adds a free function that does basic zim file validations. This function can be called by an external tool (example zimcheck).

More info: https://github.com/openzim/libzim/issues/359

### CODE REVIEW

**Comment by kelson42** [comment]
@MiguelRocha Any news here?

**Comment by MiguelRocha** [comment]
@mgautierfr can you please review?

File: src/file.cpp [hunk 1]
```diff
@@ -309,4 +309,35 @@ namespace zim
...
+          zim::File f(zimPath);
+
+          auto nrArticles = f.getCountArticles();
+
+          urlList.reserve(nrArticles);
+          titleList.reserve(nrArticles);
+
+          for (zim::article_index_type i = 0; i < nrArticles; i++)
+          {
+              auto article = f.getArticle(i);
```
Comment (mgautierfr) [inline src/file.cpp:328]: Getting a article using its index will use the `UrlPtrList` to get the article. So the article should indeed be url sorted (or url increasing as index is increasing).

But it say nothing about the title. Even on a correct zim file, the title order may differ from the url order.
This is way we have the TitlePtrList.

- `Dirent.atOffset(UrlPtrList[idx]).fullurl` should be increasing when idx is increasing
- `Dirent.atOffset(UrlPtrList[TitlePtrList[idx]]).fulltitle` should be increasing when idx is increasing.

(`atOffset` and `fulltitle` don't exist but you got the idea :) )

**Review by mgautierfr — CHANGES_REQUESTED** [review]

**Comment by veloman-yunkan** [comment]
This version covers the first step from https://github.com/openzim/libzim/issues/359#issuecomment-686373390:

> The fist step is to create this function and do some easy checks in it :
> 
> - Magic number in zim header is valid.
> - Offsets in zim header are valid (< zim size).
> - Offsets in UrlPtrList and ClusterPtrList are valid
> - Index in TitlePtrList are valid (< articleNumber)

All checks are covered by a new unit-test `ZimFile.validate`. ZIM files used by that unit-test must be generated using the `test/data/create_test_zimfiles` script (they are version controlled and must be be updated and committed whenever the source data or the said script changes).

**Review by mgautierfr — APPROVED** [review]
This PR version doesn't test the dirent content (while previous version does).
It means that it doesn't test that mimetype is correct or that url and title are correctly sorted.

However, I'm good with this PR. It introduce nice validity structure we can extend with mimetype/url/title check later.

**Comment by kelson42** [comment]
@veloman-yunkan @mgautierfr What is left here to do?

**Comment by mgautierfr** [comment]
> What is left here to do?

Technically nothing. The code is good.
Functionally, there is no check that MimeTypes are valid or UrlPtrList and TitlePtrList are correctly sorted.

I propose to merge this PR as it is and add missing tests in another PR.

**Pull request merged**